### ISSUE TITLE
agent memory balloons when unable to retrieve consul credential for template

### ISSUE THREAD

**Issue by dekimsey** [hashicorp/vault#13589]
**Describe the bug**
I'm deploying a vault-agent with aws auto_auth to generate some templated files. Due to a misconfiguration, the agent is unable to obtain the consul credential. It attempts to retry in extremely tight loop until it runs itself out of memory.

(The consul policy wasn't applied, or the vault addr isn't being applied correctly not sure yet)

**To Reproduce**
1. Run `vault agent`
2. Machine dies

This is being deployed via the Hashicorp vault rpm. To make it work as an agent I added the following systemd override:
```
[Service]
ExecStart=
ExecStart=/usr/bin/vault agent -config=/etc/vault.d/vault.hcl
# To prevent the box from dying
MemoryLimit=1G
```

**Expected behavior**
I expect the retry loop, just not running the box out of memory.

**Environment:**
* Vault Server Version (retrieve with `vault status`): v1.9.1
* Vault CLI Version (retrieve with `vault version`): v1.9.2
* Server Operating System/Architecture: CentOS 7 x86_64

Vault server configuration file(s):

```hcl
auto_auth {
  method "aws" {
    mount_path = "auth/aws"
    config = {
      type = "iam"
      role = "redacted-role-name"
      header_value = "vault.example.com"
    }
  }
  sink "file" {
    config = {
      path = "/tmp/sink"
    }
  }
}
cache {
  use_auto_auth_token   = true
  force_auto_auth_token = true
}
listener "tcp" {
  address     = "127.0.0.1:8100"
  tls_disable = true
}
template {
    destination          = "/etc/consul.d/consul.env"
    error_on_missing_key = true
    contents             = <<EOF
{{ with secret "consul/creds/redacted-role-name" }}{{.Data.token}}{{ end }}
EOF
}
template {
    destination          = "/etc/consul.d/secrets.json"
    error_on_missing_key = true
    contents             = <<EOF
{{ with secret "consul/creds/redacted-role-name" }}
{{scratch.Set "default_token" .Data.token}}
{{ end }}
{{ with secret "consul/creds/agent" }}
{{scratch.Set "agent_token" .Data.token}}
{{ end }}
{{ with secret "secret/consul/agent" }}
{{scratch.Set "encrypt" .Data.data.gossip_encryption_key }}
{{ end }}
{
  "encrypt": "{{scratch.Get "encrypt"}}",
  "acl": {
      "tokens": {
          "default": "{{scratch.Get "default_token"}}",
          "agent": "{{scratch.Get "agent_token"}}"
      }
  },
  "ca_file": "/etc/consul.d/consul-ca.pem"
}
EOF
}
template {
    destination          = "/etc/consul.d/consul-ca.pem"
    error_on_missing_key = true
    contents             = <<EOF
{{ with secret "secret/consul/agent" }}
{{ .Data.data.consul_ca }}
{{ end }}
EOF
}
```

**Additional context**
Interestingly it seems there are two configured vault addresses in my logs. The one dumped in the debug, and the one I set via environment variables which the logs show attempts at accessing _both_.

Complete log from a single process execution available here: https://gist.github.com/dekimsey/6baf8008b73c16c8a8a2f6c47ff87ef8

/etc/vault.d/vault.env:
```
VAULT_ADDR=https://vault.example.com:443
VAULT_NAMESPACE=
VAULT_LOG_LEVEL=debug
```

**Comment by hsimon-hashicorp** [comment]
My suspicion is that this is probably also related to #12566, because the Agent isn't doing retry backoffs appropriately.

**Comment by dekimsey** [comment]
@hsimon-hashicorp You are ninja-fast on triaging!

**Comment by dekimsey** [comment]
Looks like #13363 also explains the weirdness with the VAULT_ADDR environment variable seemingly not quite taking effect.

**Comment by hsimon-hashicorp** [comment]
> Looks like #13363 also explains the weirdness with the VAULT_ADDR environment variable seemingly not quite taking effect.

Indeed! Good catch. That helps a lot, in providing us more data. Thanks! :)

**Comment by dekimsey** [comment]
It does look like 1.8.7 doesn't exhibit quite the same behavior. Both the VAULT_ADDR seems to stick properly and the tight loop doesn't break the server.

**Comment by calvn** [comment]
@dekimsey can you post the logs for the 1.8.7 run if you still have that handy?

**Comment by dekimsey** [comment]
@calvn Sure thing, lemme try and get those

**Comment by dekimsey** [comment]
Okay, I updated the gist in the original report to include the 1.8.7 logs.

**Comment by calvn** [comment]
@dekimsey Thanks for that! It looks like in both runs there were permission issues. Did you observe ballooning memory usage on the 1.8.7 run as well?

**Comment by dekimsey** [comment]
@calvn Nope, 1.8.7 doesn't exhibit the same behavior.

I do see a slow climb starting from around 38M. But I don't know if that's just GC behavior. I haven't watched it long enough to verify.

**Comment by calvn** [comment]
Gotcha, thanks for sharing that info. We'd probably have to dig a bit further to see where the memory leak is occurring. 

The bit of good news is that we think the 127.0.0.1 address that you're seeing in 1.9.2 is more of a visual bug. We're now routing templating requests through Agent's cache via `bufconn`, but still need to [pass a dummy value to the dialer](https://github.com/hashicorp/vault/blob/09d6d47503b93fe899ed4a4199ca8770e4603068/command/agent/template/template.go#L293) so that it gets honored properly.

**Comment by dekimsey** [comment]
Observations on 1.8.7 overnight. The agent stabilized to 192M. So while memory usage might be higher than I'd expect for what it's doing, it's certainly reasonable and stable given garbage collection mechanics and all.

**Comment by joecorall** [comment]
I was running into this same issue and upgrading to Vault CLI v1.9.4 solved it for me.

**Comment by fairclothjm** [comment]
@dekimsey The following vault agent config setting will prevent vault agent from getting stuck in a loop

```
template_config {
  exit_on_retry_failure = true
}
```

see: https://www.vaultproject.io/docs/agent/template-config

**Comment by VioletHynes** [comment]
Hi there! I believe this will be fixed by https://github.com/hashicorp/vault/pull/16970 - if you have caching in the Vault Agent config, currently, Agent will ignore the retry configuration and perform immediate infinite retries in a loop with no back-off. As a workaround, an empty `template_config` stanza should work, by setting retry to the default config (12), though obviously that's not perfect as you might not want 12.

The fix should be coming in 1.12.

**Comment by VioletHynes** [comment]
Hi there! I'm going to close this issue as I just merged https://github.com/hashicorp/vault/pull/16970 which should fix this. This should release in 1.12.

Thanks for the bug report!
